### ISSUE TITLE
[ER-59] E2EE: Send media keys via to_device message

### ISSUE THREAD

**Issue by dbkr** [vector-im/element-call#1281]
# Add to-device key distribution

- After https://github.com/vector-im/element-call/issues/1278

In addition to sending media keys as events to the room, send them via to_device events encrypted with Olm as a primary way of sending the keys, then keep the room events as backup. Add a posthog event to track when the keys don't arrive via to_device. This will allow us to migrate away from using room events in the long term as they cause the media key to be saved in the room history.

## Prerequisits:
 - [x] To device widget MSC https://github.com/matrix-org/matrix-spec-proposals/pull/3819
 - [x] Make all platforms support to-device widget features
    - [x]  rust sdk https://github.com/matrix-org/matrix-rust-sdk/issues/4859
    - [x]   js-sdk (matrix-widget-api) https://github.com/matrix-org/matrix-widget-api/pull/57
 - [x] split the encryption key logic out from MatrixRTC session into an EncryptionManager (Done in: https://github.com/matrix-org/matrix-js-sdk/pull/4612)
 - [x]  Planning meeting what is needed and how it should be architectured (in regards to https://github.com/matrix-org/matrix-js-sdk/pull/4612) (results of planning meeting see **Production ready requirements** and **Tasks**

## Production ready requirements
 - [x] ~~Exclude untrusted devices (expected to need widget api MSC changes)~~ removed, now part of [ER-245](https://github.com/element-hq/voip-internal/issues/356)
 - [x] ~~Error handling on sending to-device events (expected to need widget api MSC changes)~~ removed, now part of [ER-244](https://github.com/element-hq/voip-internal/issues/355)
 - [x] ~~Performance: Investigate, test the performance and eventually add ratcheting (expected to need matrixRTC MSC changes)~~ removed, now part of [ER-244](https://github.com/element-hq/voip-internal/issues/355)

## Tasks

Iterations: are based on Temp: see: https://excalidraw.com/#room=b7556403b8947d5ee7ac,duF48bj2SbmRTsa3ivmLFA

### Iteration 1: "change transport" and "make it backwards compatible"
- [x] Use hughs work as reference
- [x] https://github.com/matrix-org/matrix-js-sdk/pull/4773
   - [x] move transport in its own class
   - [x] create a new transport for room event key sharing
   - [x] use that transport
   - [x] adjust tests to support the new architecture
- [x] #3155
   - [x] https://github.com/matrix-org/matrix-js-sdk/pull/4785 
      - [x] add to device Transport
      - [x] real world test
      - [x] rebase on top of https://github.com/matrix-org/matrix-js-sdk/pull/4773
      -  ~~add benchmarking (might be another PR)~~ Not part of Iteration 1, see iteration 2
      - [x] add `sendAndEncryptToDevice` message for embedded and stand alone
      - [x] test with rust sdk
      - [x] https://github.com/matrix-org/matrix-js-sdk/pull/4788
- [x] https://github.com/element-hq/element-call/pull/3167
   - [x] Do KISS backwards compatible system (just a bool flag in EC dev tools)
   - [x] use the correct js-sdk version

- [x] Discuss how we do the transition with backwards compatibility
   - Support to-device and room encryption in parallel or 
   - With smart switching (Decision: this is the right solution)
     - [x] https://github.com/matrix-org/matrix-js-sdk/pull/4792
     - [x] https://github.com/element-hq/element-call/pull/3185

### ~~Iteration 2: "to device message format"~~ [ER-244](https://github.com/element-hq/voip-internal/issues/355)
- Encrypt and send duration logging to gather some stats on N^2 scaling
- Make to device messages spec conformant: [ToDevice specification part of MatrixRTC spec 4143](https://github.com/matrix-org/matrix-spec-proposals/blob/toger5/matrixRTC/proposals/4143-matrix-rtc.md)
  - to-device will use the new format
  - room event transport will keep the old format

### ~~Iteration 3: "encryption error cases and security"~~ [ER-244](https://github.com/element-hq/voip-internal/issues/355)
-  https://github.com/matrix-org/matrix-js-sdk/pull/4792https://github.com/matrix-org/matrix-js-sdk/pull/4793 Allow the js-sdk to return errors if todevice sending failed.
-  Introduce error cases for encryption to be able to imply appropritate retry logic
-  Solve "not send to insecure devices" UX.

### ~~Other Parallelize tasks (__optional__)~~ [ER-244](https://github.com/element-hq/voip-internal/issues/355)
  - [x] Expose encryption information (`DecryptionInfo`) of to device messages. This is needed to let the widget know if a to-device message can be trusted. Is save in terms of: signed, not leaked to the HS in plaintext ... (maybe there is workaround)  https://github.com/matrix-org/matrix-rust-sdk/pull/5074

  -  Refactor the `EncryptionManager` https://github.com/matrix-org/matrix-js-sdk/pull/4799
    - [x] Rethink when we send updated new keys to minimize traffic and "time where post compromising is not fulfilled"
     - Test livekit ratcheting: https://github.com/matrix-org/matrix-js-sdk/pull/4807
     - Add option to livekit to expose allow extracting ratcheted keys. (js-client livekit client repo)
     - refactor retrying to be part of the room-transport instead of the encryption manager (update `BasicEncryptionManager` and `EncryptionManager`) @BillCarsonFr 
     - rename `BasicEncryptionManger` into `NewEncryptionManager` (maybe?)
  - Playwright key sharing tests
     - performance metrics for key sharing: Do we need ratchiting

 - [x] Prototype: 
   - [x] https://github.com/element-hq/element-call/pull/2572
   - [x] https://github.com/matrix-org/matrix-js-sdk/pull/4485

[ER-245]: https://element-io.atlassian.net/browse/ER-245?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by BillCarsonFr** [comment]
Update on the on the task for **Make all platforms support to-device widget features** 

- Rust SDK support
     - [x] Support for Clear to-device
         - https://github.com/matrix-org/matrix-rust-sdk/pull/4963
     - [x] Support for encrypted to-device
         - [x] Receiving:
             - https://github.com/matrix-org/matrix-rust-sdk/pull/4935
             - https://github.com/matrix-org/matrix-rust-sdk/pull/5074
             - https://github.com/element-hq/voip-internal/issues/347 fixed by https://github.com/matrix-org/matrix-rust-sdk/pull/5099
             - https://github.com/element-hq/voip-internal/issues/348
         - [x] Sending:
             - https://github.com/matrix-org/matrix-rust-sdk/pull/4998
             - https://github.com/element-hq/voip-internal/issues/349
                 - https://github.com/matrix-org/matrix-rust-sdk/pull/5098

- Spec Changes
    - https://github.com/element-hq/voip-internal/issues/350 Do not use the `encrypted` boolean thing, use the room encryption state (and encryption preferencese)
        - Update the MSC
        - Update rust implem
        - Update js-sdk implem

**Comment by toger5** [comment]
@fkwp
@BillCarsonFr and @toger5 went through this and concluded this can get closed.

**Comment by toger5** [comment]
Closing due to no objection to previous conclusion.
